### ISSUE TITLE
[B1.3] Implement AdaGrad optimizer in shared/autograd/optimizers.mojo

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2300]
## Overview
Implement the AdaGrad optimizer for adaptive learning rates.

## Location
- `shared/autograd/optimizers.mojo` (TODO already exists at line 157)

## Algorithm
AdaGrad adapts learning rate based on accumulated squared gradients:
```
G_t = G_{t-1} + g_t²                  # Accumulated squared gradients
θ_t = θ_{t-1} - α * g_t / (√G_t + ε)  # Update
```

## Deliverables
- [ ] Implement AdaGrad struct with:
  - learning_rate (default: 0.01)
  - epsilon (default: 1e-10)
  - weight_decay (default: 0.0)
- [ ] Implement step() method matching SGD interface
- [ ] Implement zero_grad() method
- [ ] Store accumulated squared gradient buffers
- [ ] Export from `shared/autograd/__init__.mojo`

## Success Criteria
- AdaGrad optimizer works with Variable/GradientTape system
- Training converges on test problems
- CI passes

## Dependencies
- None (parallel with B1.1, B1.2)

## Track
Track B: Missing Features (B1.3 of 8)
